docs/spec/constructors.rst

```diff
@@ -0,0 +1,375 @@
+parameter type is explicit or inferred, a type checker should bind the class
+being constructed to this parameter and report any type errors that arise
+during binding.
+
+  ::
+
+    class MyClass[T]:
+        def __init__(self: "MyClass[int]") -> "None": ...
+
+    MyClass()  # OK
+    MyClass[int]()  # OK
+    MyClass[str]()  # Type Error
+
+The return type for ``__init__`` is always ``None``, which means the
+method cannot influence the return type of the constructor call by specifying
+a return type. To work around this limitation, type checkers should allow
+the ``self`` parameter to be annotated with a type that influences the resulting
+type of the constructor call. This can be used in overloads to influence the
+constructor return type for each overload.
+
+  ::
+
+    class MyClass1[T]:
+        @overload
+        def __init__(self: "MyClass1[list[int]]", value: int) -> None: ...
+        @overload
+        def __init__(self: "MyClass1[set[str]]", value: str) -> None: ...
+        @overload
+        def __init__(self, value: T) -> None: ...
+
+
+    assert_type(MyClass1(0), MyClass1[list[int]])
+    assert_type(MyClass1[int](3), MyClass1[int])
+    assert_type(MyClass1(""), MyClass1[set[str]])
+    assert_type(MyClass1(3.0), MyClass1[float])
+
+
+Function-scoped type variables can also be used in the ``self``
+annotation of an ``__init__`` method to influence the return type of the
+constructor call.
+
+  ::
+
+    class MyClass2[T1, T2]:
+        def __init__[V1, V2](self: "MyClass2[V1, V2]", value1: V1, value2: V2) -> None: ...
+
+    assert_type(MyClass2(0, ""), MyClass2[int, str])
+    assert_type(MyClass2[int, str](0, ""), MyClass2[int, str])
+
+    class MyClass3[T1, T2]:
+        def __init__[V1, V2](self: "MyClass3[V2, V1]", value1: V1, value2: V2) -> None: ...
+
+    assert_type(MyClass3(0, ""), MyClass3[str, int])
+    assert_type(MyClass3[str, int](0, ""), MyClass3[str, int])
+
+
+Class-scoped type variables should not be used in the ``self`` annotation
+because such use can lead to ambiguous or nonsensical type evaluation results.
+Type checkers should report an error if a class-scoped type variable is used
+within a type annotation for the ``self`` parameter in an ``__init__`` method.
+
+  ::
+
+    class MyClass4[T1, T2]:
+        # The ``self`` annotation should result in a type error
+        def __init__(self: "MyClass4[T2, T1]") -> None: ...
+
+
+Constructor Calls for type[T]
+-----------------------------
+
+When a value of type ``type[T]`` (where ``T`` is a type variable or a concrete
+class) is called, a type checker should evaluate the constructor call as if
+it is being made on the class ``T`` (or the class the represents the upper bound
+of type variable ``T``). This means the type checker should use ``__call__``
+method of ``T``'s metaclass and the ``__new__`` and ``__init__`` methods of ``T``
+to evaluate the constructor call.
+
+It should be noted that such code could be unsafe because The type ``type[T]``
```

Comment:
```suggestion
It should be noted that such code could be unsafe because the type ``type[T]``
```
This paragraph in general feels a little out of place because there is no normative suggestion, only a description.